stable/database/files/backup_hooks.py

```diff
@@ -0,0 +1,386 @@
+                    state = line[len(prefix) :].strip()
+                    if state != "T (stopped)":
+                        LOGGER.info("Thread %s not stopped: %s", tid, state)
+                        return False
+                    return True
+    except FileNotFoundError as e:
+        LOGGER.warning("Unable to get state of thread %s: %s", tid, e)
+
+
+def check_suspended(pid):
+    # Check if any threads are not in suspended (stopped) state
+    for tid in os.listdir(f"/proc/{pid}/task"):
+        if tid.isdigit() and not check_thread_suspended(pid, tid):
+            return False
+    return True
+
+
+def await_suspended(pid, interval=0.25, retries=16):
+    for i in range(0, retries):
+        if i == 0:
+            LOGGER.info("Waiting for all threads of process %s to be suspended", pid)
+        else:
+            LOGGER.info(
+                "Retrying in %s second(s) (iteration %s/%s)", interval, i + 1, retries
+            )
+            time.sleep(interval)
+
+        # If all threads are suspended, return immediately
+        if check_suspended(pid):
+            return
+
+    raise RuntimeError(
+        "Process {} not suspended after {} seconds".format(pid, retries * interval)
+    )
+
+
+def freeze_archive(unfreeze=False):
+    if USE_SUSPEND:
+        # Resume or suspend nuodb process. There should be a unique nuodb
+        # process, but if there are multiple somehow, suspend all of them.
+        pids = get_nuodb_pids()
+        if not pids:
+            raise RuntimeError("No nuodb process found")
+        for pid in pids:
+            if unfreeze:
+                LOGGER.info("Resuming nuodb process with pid %s", pid)
+                os.kill(int(pid), signal.SIGCONT)
+            else:
+                LOGGER.info("Suspending nuodb process with pid %s", pid)
+                os.kill(int(pid), signal.SIGSTOP)
+                # Make sure all threads are suspended
+                await_suspended(pid)
+    else:
+        # Freeze or unfreeze the archive filesystem using fsfreeze
+        if unfreeze:
+            LOGGER.info("Unfreezing writes to archive volume")
+            action = "--unfreeze"
+        else:
+            LOGGER.info("Freezing writes to archive volume")
+            action = "--freeze"
+        subprocess.check_output(
+            ["fsfreeze", action, ARCHIVE_DIR], stderr=subprocess.STDOUT
+        )
+
+
+def pre_backup(backup_id, payload):
+    # Make sure we do not attempt to execute pre-backup hook if a backup is
+    # already in progress, which may block indefinitely if writes to the
+    # archive and journal directories are frozen
+    if os.path.exists(ARCHIVE_BACKUP_ID_FILE):
+        raise UserError(
+            "Backup ID file {} exists. Execute post-backup hook to complete current backup.".format(
+                ARCHIVE_BACKUP_ID_FILE
+            )
+        )
+
+    # Write backup ID to archive directory
+    write_file(ARCHIVE_BACKUP_ID_FILE, backup_id)
```

Comment:
Should we validate that the supplied {{backup_id}} is valid (i.e not `None`)?

Reply:
Good idea. The request dispatcher checks that the correct number of path components are specified, but the user can still supply an empty backup ID which we should prevent.